core/src/main/java/org/infinispan/persistence/PersistenceUtil.java

```diff
@@ -89,7 +89,7 @@ public static InternalMetadata internalMetadata(InternalCacheValue icv) {
                                                          final AtomicReference<Boolean> isLoaded) {
       return dataContainer.compute(key, (k, oldEntry, factory) -> {
          //under the lock, check if the entry exists in the DataContainer
-         if (oldEntry != null) {
+         if (oldEntry != null && (!oldEntry.canExpire() || !oldEntry.isExpired(timeService.wallClockTime()))) {
```

Comment:
Unfortunately a lot of places call load without invoking this method :(  The big problem is this wouldn't call to load if it found an expired entry.